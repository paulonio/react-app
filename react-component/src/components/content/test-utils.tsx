import React, { FC, ReactElement, useReducer } from 'react';
import { initialState, mainReducer, SearchContext } from 'context/context';

export const StateProvider: FC<{ children: React.ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(mainReducer, initialState);
  return <SearchContext.Provider value={{ state, dispatch }}>{children}</SearchContext.Provider>;
};
