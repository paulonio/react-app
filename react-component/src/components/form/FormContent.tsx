import React, { useContext, useEffect } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';
import FormCards from './FormCards';
import classes from './formContent.module.scss';
import MyBirthDateInput from 'components/form/form-inputs/MyBirthDateInput';
import MyFileInput from 'components/form/form-inputs/MyFileInput';
import MySelectInput from 'components/form/form-inputs/MySelectInput';
import MySwitcherInput from 'components/form/form-inputs/MySwitcherInput';
import MyCheckboxInput from 'components/form/form-inputs/MyCheckboxInput';
import MyInputName from 'components/form/form-inputs/MyInputName';
import MyInputLastName from 'components/form/form-inputs/MyInputLastName';
import MyFormButtons from 'components/form/form-inputs/MyFormButtons';
import { SearchContext } from 'context/context';
import { Action } from 'context/reducers';

export interface IFormData {
  name: string;
  lastName: string;
  birthDate: string;
  avatar: File[] | string;
  country: string;
  personalData: boolean;
  notifications: boolean;
}

const FormContent = () => {
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid, isSubmitted, isSubmitSuccessful },
    getValues,
  } = useForm<IFormData>({ mode: 'onSubmit', reValidateMode: 'onChange' });
  const { state, dispatch } = useContext(SearchContext);

  useEffect(() => {
    if (isSubmitSuccessful) {
      reset();
      reset({ personalData: false, notifications: false });
    }
  }, [isSubmitSuccessful, reset]);

  const onSubmit: SubmitHandler<IFormData> = (data) => {
    const avatarURL = URL.createObjectURL(getValues('avatar')[0] as File);
    data.avatar = avatarURL;
    const action: Action = { type: 'UPDATE_FORM_DATA', data };
    dispatch(action);
  };

  return (
    <>
      <section className={classes.form__content}>
        <h1 className={classes.title}>Form</h1>
        <form className={classes.form__container} onSubmit={handleSubmit(onSubmit)}>
          <MyInputName register={register} error={errors.name} />
          <MyInputLastName register={register} error={errors.lastName} />
          <MyBirthDateInput register={register} error={errors.birthDate} />
          <MyFileInput register={register} error={errors.avatar} />
          <MySelectInput register={register} error={errors.country} />
          <MySwitcherInput register={register} error={errors.personalData} />
          <MyCheckboxInput register={register} error={errors.notifications} />
          <MyFormButtons isDisabled={isSubmitted ? !isValid : !isDirty} />
        </form>
      </section>
      <FormCards data={state.formData} />
    </>
  );
};

export default FormContent;
