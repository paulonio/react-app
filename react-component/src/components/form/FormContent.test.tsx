import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import FormContent from './FormContent';
import { act } from 'react-dom/test-utils';

describe('Content', () => {
  it('should render form', () => {
    const { container } = render(<FormContent />);
    expect(container).toMatchSnapshot();
  });
  it('should set value to name input', async () => {
    render(<FormContent />);
    const nameInput = screen.getByPlaceholderText(/type your name/i) as HTMLInputElement;
    const submitButton = screen.getByText(/submit/i);
    const testValue = 'John';
    await act(() => {
      fireEvent.change(nameInput, { target: { value: testValue } });
      fireEvent.click(submitButton);
    });
    expect(nameInput.value).toEqual(testValue);
  });
  it('should throw error in case of invalid name', async () => {
    render(<FormContent />);
    const nameInput = screen.getByPlaceholderText(/type your name/i) as HTMLInputElement;
    const submitButton = screen.getByText(/submit/i);
    const fakeValue = '1234';
    fireEvent.change(nameInput, { target: { value: fakeValue } });
    fireEvent.click(submitButton);
    await waitFor(() => {
      expect(screen.getByText('The field should contain only letters.')).toBeInTheDocument();
    });
  });
  it('should set date to date input', async () => {
    render(<FormContent />);
    const dateInput = screen.getByLabelText(/birth date/i) as HTMLInputElement;
    const submitButton = screen.getByText(/submit/i);
    const date = '2022-09-27';
    await act(() => {
      fireEvent.change(dateInput, { target: { value: date } });
      fireEvent.click(submitButton);
    });
    expect(dateInput.value).toEqual(date);
  });
  it('should throw error in case of invalid date', async () => {
    render(<FormContent />);
    const dateInput = screen.getByLabelText(/birth date/i) as HTMLInputElement;
    const submitButton = screen.getByText(/submit/i);
    const date = '2045-09-27';
    fireEvent.change(dateInput, { target: { value: date } });
    fireEvent.click(submitButton);
    await waitFor(() => {
      expect(screen.getByText(/select a valid date/i)).toBeInTheDocument();
    });
  });
  it('should not throw an error in case of checked switcher', () => {
    const { container } = render(<FormContent />);
    const switcher = container.querySelector('.input__switcher') as HTMLInputElement;
    expect(switcher.checked).toEqual(false);
    fireEvent.click(switcher);
    expect(switcher.checked).toEqual(true);
  });
  it('should not throw an error in case of checked checkbox', () => {
    const { container } = render(<FormContent />);
    const checkbox = container.querySelector('.input__checkbox') as HTMLInputElement;
    expect(checkbox.checked).toEqual(false);
    fireEvent.click(checkbox);
    expect(checkbox.checked).toEqual(true);
  });
  it('should enable submit button on form changing', () => {
    const { container } = render(<FormContent />);
    const nameInput = container.querySelector('.text__input') as HTMLInputElement;
    const submitButton = screen.getByText(/submit/i) as HTMLButtonElement;
    const testString = 'John';
    expect(submitButton.disabled).toEqual(true);
    fireEvent.change(nameInput, { target: { value: testString } });
    expect(submitButton.disabled).toEqual(false);
  });
});
