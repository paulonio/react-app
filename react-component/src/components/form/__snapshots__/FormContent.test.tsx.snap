// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content should render form 1`] = `
<div>
  <section
    class="form__content"
  >
    <h1
      class="title"
    >
      Form
    </h1>
    <form
      class="form__container"
    >
      <label
        class="label"
      >
        <div
          class="input__title"
        >
          Name
        </div>
        <input
          class="text__input"
          name="name"
          placeholder="Type your name"
          type="text"
        />
      </label>
      <div
        class="error__message"
      />
      <label
        class="label"
      >
        <div
          class="input__title"
        >
          Last name
        </div>
        <input
          class="text__input"
          name="lastName"
          placeholder="Type your last name"
          type="text"
        />
      </label>
      <div
        class="error__message"
      />
      <label
        class="label"
      >
        <div
          class="input__title"
        >
          Birth date
        </div>
        <input
          class="input__birth_date"
          name="birthDate"
          type="date"
        />
      </label>
      <div
        class="error__message"
      />
      <label
        class="label"
      >
        <div
          class="input__title"
        >
          Avatar
        </div>
        <input
          accept="image/png, image/jpeg"
          class="input__avatar"
          name="avatar"
          type="file"
        />
      </label>
      <div
        class="error__message"
      />
      <label
        class="label"
      >
        <div
          class="input__title"
        >
          Choose your country
        </div>
        <select
          class="select__country"
          name="country"
        >
          <option
            value="Belarus"
          >
            Belarus
          </option>
          <option
            value="Russia"
          >
            Russia
          </option>
          <option
            value="Ukraine"
          >
            Ukraine
          </option>
          <option
            value="USA"
          >
            USA
          </option>
        </select>
      </label>
      <div
        class="error__message"
      />
      <label
        class="label__personal_data"
      >
        <input
          class="input__switcher"
          name="personalData"
          type="checkbox"
        />
        <span
          class="switcher"
        />
        <div
          class="personal_data"
        >
          I agree that my personal data will be processed
        </div>
      </label>
      <div
        class="error__message"
      />
      <label
        class="label__notifications"
      >
        <input
          class="input__checkbox"
          name="notifications"
          type="checkbox"
        />
        <div
          class="notifications"
        >
          I agree to receive notifications
        </div>
      </label>
      <div
        class="error__message"
      />
      <div
        class="form__buttons"
      >
        <button
          class="button"
          disabled=""
          type="submit"
        >
          Submit
        </button>
        <button
          class="button"
          type="reset"
        >
          Reset
        </button>
      </div>
    </form>
  </section>
  <section
    class="form__cards"
  />
</div>
`;
